build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/release/UpdateBranchesJsonTask.java

```diff
@@ -0,0 +1,152 @@
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+import org.elasticsearch.gradle.Version;
+import org.elasticsearch.gradle.internal.conventions.util.Util;
+import org.elasticsearch.gradle.internal.info.BranchesFileParser;
+import org.elasticsearch.gradle.internal.info.DevelopmentBranch;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.InvalidUserDataException;
+import org.gradle.api.Project;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.options.Option;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.util.Comparator;
+import java.util.List;
+
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+
+/**
+ * Updates the branches.json file in the root of the repository
+ */
+public class UpdateBranchesJsonTask extends DefaultTask {
+
+    private static final Logger LOGGER = Logging.getLogger(UpdateBranchesJsonTask.class);
+
+    private final Project project;
+    private final ObjectMapper objectMapper;
+    private final BranchesFileParser branchesFileParser;
+
+    @Nullable
+    private DevelopmentBranch addBranch;
+    @Nullable
+    private String removeBranch;
+    @Nullable
+    private DevelopmentBranch updateBranch;
+
+    @Inject
+    public UpdateBranchesJsonTask(Project project) {
+        this.project = project;
+        this.objectMapper = new ObjectMapper();
+        this.branchesFileParser = new BranchesFileParser(objectMapper);
+    }
+
+    @Option(option = "add-branch", description = "Specifies the branch and corresponding version to add in format <branch>:<version>")
+    public void addBranch(String branchAndVersion) {
+        this.addBranch = toDevelopmentBranch(branchAndVersion);
+    }
+
+    @Option(option = "remove-branch", description = "Specifies the branch to remove")
+    public void removeBranch(String branch) {
+        this.removeBranch = branch;
+    }
+
+    @Option(option = "update-branch", description = "Specifies the branch and corresponding version to update in format <branch>:<version>")
+    public void updateBranch(String branchAndVersion) {
+        this.updateBranch = toDevelopmentBranch(branchAndVersion);
+    }
+
+    private DevelopmentBranch toDevelopmentBranch(String branchAndVersion) {
+        String[] parts = branchAndVersion.split(":");
+        if (parts.length != 2) {
+            throw new InvalidUserDataException("Expected branch and version in format <branch>:<version>");
+        }
+        return new DevelopmentBranch(parts[0], Version.fromString(parts[1]));
+    }
+
+    @TaskAction
+    public void executeTask() throws IOException {
+        File branchesFile = new File(Util.locateElasticsearchWorkspace(project.getGradle()), "branches.json");
```

Comment:
referencing `project` here while executing task is deprecated. In this case I think we should just inject `ProjectLayout` instead of project in the ctor and then use `layout.getSettingsDirectory().file('branches.json').asFile`. 

The `Util.locateElasticsearchWorkspace` is just a utility method only used to resolve files relateive from internal builds within this repository, which we don't need here as we never run this task anywhere else but in the :elasticsearch root build

Reply:
Also we probably want to make this file an outputFile property of the task